http/src/v0/root_files.rs

```diff
@@ -58,3 +68,373 @@ async fn cat_inner<T: IpfsTypes>(ipfs: Ipfs<T>, args: CatArgs) -> Result<impl Re
+
+                    if segment.is_first() {
+                        let path = item.as_entry().path();
+                        let metadata = item
+                            .as_entry()
+                            .metadata()
+                            .expect("files must have metadata");
+
+                        for mut bytes in tar_helper.apply_file(path, metadata, total_size)?.iter_mut() {
+                            if let Some(bytes) = bytes.take() {
+                                yield bytes;
+                            }
+                        }
+                    }
+
+                    // even if the largest of files can have 256 kB blocks and about the same
+                    // amount of content, try to consume it in small parts not to grow the buffers
+                    // too much.
+
+                    let mut n = 0usize;
+                    let slice = segment.as_ref();
+                    let total = slice.len();
+
+                    while n < total {
+                        let next = tar_helper.buffer_file_contents(&slice[n..]);
+                        n += next.len();
+                        yield next;
+                    }
+
+                    if segment.is_last() {
+                        if let Some(zeroes) = tar_helper.pad(total_size) {
+                            yield zeroes;
+                        }
+                    }
+
+                    item.into_inner()
+                },
+                ContinuedWalk::Directory(item) => {
+
+                    // only first instances of directorys will have the metadata
+                    if let Some(metadata) = item.as_entry().metadata() {
+                        let path = item.as_entry().path();
+
+                        for mut bytes in tar_helper.apply_directory(path, metadata)?.iter_mut() {
+                            if let Some(bytes) = bytes.take() {
+                                yield bytes;
+                            }
+                        }
+                    }
+
+                    item.into_inner()
+                },
+                ContinuedWalk::Symlink(bytes, item) => {
+
+                    // converting a symlink is the most tricky part
+                    let path = item.as_entry().path();
+                    let target = std::str::from_utf8(bytes).map_err(|_| GetError::NonUtf8Symlink)?;
+                    let target = Path::new(target);
+                    let metadata = item.as_entry().metadata().expect("symlink must have metadata");
+
+                    for mut bytes in tar_helper.apply_symlink(path, target, metadata)?.iter_mut() {
+                        if let Some(bytes) = bytes.take() {
+                            yield bytes;
+                        }
+                    }
+
+                    item.into_inner()
+                },
+            };
+        }
+    }
+}
+
+#[derive(Debug)]
+enum GetError {
+    NonUtf8Symlink,
```

Comment:
Wouldn't we want to provide the underlying bytes of a symlink that is not UTF-8? It might be "mangled" in a way that is not visible to the user, so if this is not immediately thrown only when given a direct symlink, it would be a good idea to indicate where the issue is.

Reply:
This error will actually be eaten by `hyper` as it currently doesn't support returning trailer headers. No one would ever see the bytes. The `tar` functionality is not exposed in the `ipfs` crate. This is by design, we currently can't do it using current `tar-rs` interfaces.

Reply:
The "design" idea was that we could provide the means to `get` to local filesystem, which we almost do with `ipfs::unixfs::ll::walk::Walker` (or `ipfs_unixfs::walk::Walker`).